pystiche/papers/ulyanov_et_al_2016/core.py

```diff
@@ -0,0 +1,205 @@
+            optimizer, impl_params=impl_params,
+        )
+
+    if num_epochs is None:
+        if impl_params:
+            if instance_norm:
+                num_epochs = 25
+            else:
+                num_epochs = 10 if stylization else 5
+        else:
+            num_epochs = 10
+
+    style_transform = ulyanov_et_al_2016_style_transform(
+        impl_params=impl_params, instance_norm=instance_norm
+    )
+    style_transform = style_transform.to(device)
+    preprocessor = ulyanov_et_al_2016_preprocessor()
+    preprocessor = preprocessor.to(device)
+    style_image = style_transform(style_image)
+    style_image = preprocessor(style_image)
+    style_image = batch_up_image(style_image, loader=content_image_loader)
+    criterion.set_style_image(style_image)
+
+    if stylization:
+        def criterion_update_fn(input_image, criterion):
+            criterion.set_content_image(preprocessor(input_image))
+    else:
+        def criterion_update_fn(input_image, criterion):
+            pass
+
+
+    default_transformer_epoch_optim_loop(
+        content_image_loader,
+        transformer,
+        criterion,
+        criterion_update_fn,
+        num_epochs,
+        device=device,
+        lr_scheduler=lr_scheduler,
+        quiet=quiet,
+        logger=logger,
+        log_fn=log_fn,
+    )
+
+    return transformer
+
+
+def ulyanov_et_al_2016_style_generation(
+    input_image: torch.Tensor,
+    transformer: Union[nn.Module, str],
+    impl_params: bool = True,
+    instance_norm: bool = False,
+    stylization: bool = True,
+    postprocessor: Optional[CaffePostprocessing] = None,
+):
+    device = input_image.device
+    if isinstance(transformer, str):
+        style = transformer
+        transformer = ulyanov_et_al_2016_transformer(
+            style=style, impl_params=impl_params, instance_norm=instance_norm, stylization=stylization
+        )
+        if instance_norm or not impl_params:
+            transformer = transformer.eval()
+        transformer = transformer.to(device)
+
+    with torch.no_grad():
+        content_transform = ulyanov_et_al_2016_content_transform(impl_params=impl_params,
+                                                                 instance_norm=instance_norm)
+        content_transform = content_transform.to(device)
+        input_image = content_transform(input_image)
+
+        if postprocessor is None:
+            postprocessor = ulyanov_et_al_2016_postprocessor()
+        postprocessor = postprocessor.to(device)
+        output_image = transformer(input_image)
+        output_image = postprocessor(output_image)
+
+    return output_image.detach()
+
+
+def ulyanov_et_al_2016_texture_generation(
+    input: Union[Tuple[int, int], torch.Tensor],
+    transformer: Union[nn.Module, str],
+    impl_params: bool = True,
+    instance_norm: bool = False,
+    stylization: bool = True,
+    postprocessor: Optional[CaffePostprocessing] = None,
```

Comment:
Same as above.

Reply:
This is because the images are not preprocessed for the Transformer. Neither before they go into the transformer nor after. Therefore the Transformer will transform preprocessed images and this step must be undone. There are better solutions for this but the author does it this way.

Reply:
Just checking if I understand you correctly: the `transformer` needs to learn the preprocessing during training, since it is not included as transform, correct? Thus, if we want to perform a stylization or texture generation with the `transformer` we need to postprocess the generated image, correct?

If the top paragraph is correct, I think we need to do one of two things:

1. We either include the `preprocessor` as an optional argument in `ulyanov_et_al_2016_training` or
2. we remove the optional `postprocessor` from `ulyanov_et_al_2016_style_generation` and `ulyanov_et_al_2016_texture_generation`.

We either give the user the change to explicitly set both or none. If we fix the `preprocessor` it would be nonsensical to leave the `postprocessor` up to the user, since only one will give useful results. Thoughts?

Reply:
First your top paragragh is correct. 

What advantages would it have if the user could still explicitly set both? If these do not outweigh, I would prefer variant 2 and remove the optional `postprocessor`, because this is the reprogramming of the paper and the author uses only these processors.